File: cpp/math/xor_basis.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
struct XorBasis {
  static const int LOG = 62;
  ll b[LOG + 1]{};
  int sz = 0;
  bool insert(ll x) {
    for (int i = LOG; i >= 0; i--)
      if (x >> i & 1) {
        if (!b[i]) {
          b[i] = x;
          sz++;
          return 1;
        }
        x ^= b[i];
      }
    return 0;
  }
  ll max_xor(ll x = 0) {
    for (int i = LOG; i >= 0; i--)
      x = max(x, x ^ b[i]);
    return x;
  }
  ll min_positive() {
    for (int i = 0; i <= LOG; i++)
      if (b[i])
        return b[i];
    return 0;
  }
};

File: python/math/xor_basis.py
class XorBasis:
    def __init__(self, bits=63):
        self.basis = [0] * bits

    def add(self, x):
        for i in range(len(self.basis) - 1, -1, -1):
            if not (x >> i) & 1:
                continue
            if not self.basis[i]:
                self.basis[i] = x
                return True
            x ^= self.basis[i]
        return False

    def max_xor(self, x=0):
        for b in reversed(self.basis):
            x = max(x, x ^ b)
        return x
